pkg/controller/certificates/clustertrustbundlepublisher/publisher.go

```diff
@@ -0,0 +1,365 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clustertrustbundlepublisher
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"strings"
+	"time"
+
+	certificatesv1alpha1 "k8s.io/api/certificates/v1alpha1"
+	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+	certalpha1informers "k8s.io/client-go/informers/certificates/v1alpha1"
+	certbeta1informers "k8s.io/client-go/informers/certificates/v1beta1"
+	clientset "k8s.io/client-go/kubernetes"
+	certalphav1listers "k8s.io/client-go/listers/certificates/v1alpha1"
+	certbetav1listers "k8s.io/client-go/listers/certificates/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+func init() {
+	registerMetrics()
+}
+
+type PublisherRunner interface {
+	Run(context.Context)
+}
+
+type ClusterTrustBundlePublisher[T clusterTrustBundle] struct {
+	signerName string
+	ca         dynamiccertificates.CAContentProvider
+
+	client clusterTrustBundlesClient[T]
+
+	ctbInformer     cache.SharedIndexInformer
+	ctbLister       clusterTrustBundlesLister[T]
+	ctbListerSynced cache.InformerSynced
+
+	handlers clusterTrustBundleHandlers[T]
+
+	queue workqueue.TypedRateLimitingInterface[string]
+}
+
+// clusterTrustBundle is a type constraint grouping all APIs versions of ClusterTrustBundles
+type clusterTrustBundle interface {
+	certificatesv1alpha1.ClusterTrustBundle | certificatesv1beta1.ClusterTrustBundle
```

Comment:
I don't understand why this is necessary --- why should this controller need to handle v1alpha1 ClusterTrustBundles?

Reply:
@enj thought it might be better to work with either API version

Reply:
We will definitely need it for beta -> v1 so no reason not do it here as well.